maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+            State::Empty => return WakeOutcome::NoMatch(val),
+        }
+
+        // okay, there are tasks waiting on the queue; we must acquire the lock
+        // on the linked list and wake the next task from the queue.
+        let mut queue = self.queue.lock();
+        test_trace!("wake: -> locked");
+
+        // the queue's state may have changed while we were waiting to acquire
+        // the lock, so we need to acquire a new snapshot.
+        state = self.load();
+
+        if let Some(mut node) = self.node_match_locked(key, &mut *queue, state) {
+            let waker = Waiter::<K, V>::wake(node, &mut *queue, Wakeup::DataReceived);
+            unsafe {
+                node.as_mut().val.with_mut(|v| (*v).write(val));
+            }
+            drop(queue);
+            waker.wake();
+            WakeOutcome::Woke
+        } else {
+            WakeOutcome::NoMatch(val)
+        }
+    }
+
+    /// Close the queue, indicating that it may no longer be used.
+    ///
+    /// Once a queue is closed, all [`wait`] calls (current or future) will
+    /// return an error.
+    ///
+    /// This method is generally used when implementing higher-level
+    /// synchronization primitives or resources: when an event makes a resource
+    /// permanently unavailable, the queue can be closed.
+    pub fn close(&self) {
+        let state = self.state.fetch_or(State::Closed.into_usize(), SeqCst);
+        let state = test_dbg!(State::from_bits(state));
+        if state != State::Waiting {
+            return;
+        }
+
+        let mut queue = self.queue.lock();
+
+        // TODO(eliza): wake outside the lock using an array, a la
+        // https://github.com/tokio-rs/tokio/blob/4941fbf7c43566a8f491c64af5a4cd627c99e5a6/tokio/src/sync/batch_semaphore.rs#L277-L303
+        while let Some(node) = queue.pop_back() {
+            let waker = Waiter::wake(node, &mut queue, Wakeup::Closed);
+            waker.wake()
+        }
+    }
+
+    /// Wait to be woken up by this queue.
+    ///
+    /// This returns a [`Wait`] future that will complete when the task is
+    /// woken by a call to [`wake`] with a matching `key`, or when the `WaitMap`
+    /// is dropped.
+    ///
+    /// NOTE: `key`s must be unique. If the given key already exists in the
+    /// `WaitMap`, this function will panic. See [`try_wait`] for a non-
+    /// panicking alternative.
+    ///
+    /// [`try_wait`]: Self::try_wait
+    /// [`wake`]: Self::wake
+    pub fn wait(&self, key: K) -> Wait<'_, K, V> {
+        self.try_wait(key)
+            .map_err(drop)
+            .expect("Keys should be unique in a `WaitMap`!")
+    }
+
+    /// Attempt to wait to be woken up by this queue.
+    ///
+    /// This returns a [`Wait`] future that will complete when the task is
+    /// woken by a call to [`wake`] with a matching `key`, or when the `WaitMap`
+    /// is dropped.
+    ///
+    /// NOTE: `key`s must be unique. If the given key already exists in the
```

Comment:
```suggestion
    /// **Note**: `key`s must be unique. If the given key already exists in the
```
(but i don't actually care)